### PR TITLE
Add streaming Rust implementation

### PR DESCRIPTION
**Pull request by jeffkaufman**
Port the streaming API of dedup.py to Rust. This is ~50x faster: the 1hr runtime for 2M reads I was observing with the streaming version is now just over one minute.

I originally did this in C (https://github.com/securebio/nao-dedup/pull/4) but after discussion with @evanfields and @willbradshaw we decided Rust was a better choice.  I'm much less familiar with Rust, so this relies very heavily on Claude Code: I'm not just using to to quickly write something I could write myself, but instead to translate my C and Python into a language I don't really know.

All streaming tests run against both python and rust, to avoid duplicate tests and ensure we keep implementations in sync.

I think it's good to keep both the python and rust versions of the streaming code, because organizationally we're not nearly as strong in rust and rust is less readable, so the python version does double duty as a kind of very detailed documentation of the approach.  But I could be convinced to drop the streaming python; I think there was a much stronger case for keeping both versions when I was thinking we'd do python + C.

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,381 @@
...
+// Minimizer Extraction
+// ============================================================================
+
+fn compute_kmer_hash(seq: &[u8], kmer_len: usize) -> u64 {
+    let mut hash: u64 = 0;
+    for &base in &seq[..kmer_len] {
+        hash = hash.wrapping_mul(31).wrapping_add(base as u64);
+    }
+    hash
+}
```
[the remark this answers is not in the document]
Comment (jeffkaufman) [reply src/lib.rs]: Done in 7505dab by switching to a simple k-mer rolling hash

File: src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,381 @@
...
+    // Process all reads
+    for rp in read_pairs {
+        ctx.process_read(rp);
+    }
+
+    // Finalize
+    ctx.finalize();
+
+    // Return the results (already built during processing)
+    ctx.results.clone()
```
[the remark this answers is not in the document]
Comment (jeffkaufman) [reply src/lib.rs]: Done in 7220a8a

File: src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,381 @@
...
+// ============================================================================
+
+#[derive(Debug, Clone)]
+struct ClusterStats {
+    #[allow(dead_code)]
+    key: String,           // Immutable: initial exemplar ID (used as hash key)
+    best_read_id: String,  // Mutable: current best read ID
+    best_score: f64,
+    count: usize,
+}
```
[the remark this answers is not in the document]
Comment (jeffkaufman) [reply src/lib.rs]: Fixed in 65168b6

File: tests/test_dedup.py [hunk 4]
```diff
@@ -620,3 +540,289 @@ def test_realistic_reads_with_errors(self, dedup_func):
...
+
+        result = dedup_func(read_pairs, verbose=False)
+        mapping = _get_exemplar_mapping(result)
+
+        # Each should be its own exemplar
+        for read_id in mapping:
+            assert mapping[read_id] == read_id
+
+    def test_approximate_match_parity(self, dedup_func):
+        """Ensure C and Python handle mismatches identically."""
```
[the remark this answers is not in the document]
Comment (jeffkaufman) [reply tests/test_dedup.py]: done in 2c668f3

File: Cargo.toml [hunk 5]
```diff
@@ -0,0 +1,11 @@
+[package]
+name = "nao_dedup"
+version = "0.1.0"
+edition = "2021"
```
Comment (evanfields) [inline Cargo.toml]: 2021?

File: README.md [hunk 6]
```diff
@@ -1,13 +1,23 @@
-# nao-dedup
+# nao_dedup
```
Comment (evanfields) [inline README.md]: The `README.md` has a bunch of changes which seem somewhat specialized towards the new Rust implementation rather than the 2-algorithm Python+Rust codebase as a whole. It's not obvious to me which are intentional and which are an AI coding assistant just getting overexcited about the latest PR.

Example: the repo is still called `nao-dedup`, so why rename? Probably because the Rust `Cargo.toml` in the repo root names the package `nao_dedup`?

File: src/lib.rs [hunk 7]
```diff
@@ -0,0 +1,420 @@
...
+        // Find matching exemplar by checking buckets (use FIRST match found)
+        let mut matching_exemplar: Option<String> = None;
+
+        'outer: for &min_hash in &all_mins {
+            if let Some(bucket_reads) = self.buckets.get(&min_hash) {
+                for candidate_id in bucket_reads {
+                    // Look up candidate in exemplar_store (not all reads)
+                    if let Some(candidate) = self.exemplar_store.get(candidate_id) {
+                        // Check if reads are similar
+                        if reads_are_similar(&read_pair, candidate, &self.dedup_params) {
```
Comment (evanfields) [inline src/lib.rs]: nit which I'm putting here arbitrarily, but could apply elsewhere, no required change:
```
// Check if reads are similar
if reads_are_similar(...)
```
is not a very useful comment. But overall as a new reader of Rust, comments are _very_ helpful to me in this code, and I assume likewise for the rest of the team. The verbosity (I assume AI generated) is overall nice, but on the margin fewer self-explanatory line comments and more function level "how the pieces fit together" and more "this is why we did this Rust choice" comments would be nicer.

File: README.md [hunk 8]
```diff
@@ -19,41 +29,31 @@ reads that differ slightly due to sequencing errors or alignment variations.
...
-   - **Advantages**: Uses graph centrality to select optimal exemplars
-   - **Memory usage**: Stores all reads in memory (~3GB for 250k reads)
-
-2. **`deduplicate_read_pairs_streaming()`** - Streaming two-pass algorithm
-   - **Best for**: Large datasets (> 100k reads)
-   - **Advantages**: Only stores unique sequences in memory (~2-3x less memory
-     in the typical case, far more on pathological datasets)
-   - **Memory usage**: Much lower (~1GB for 250k reads with 75k unique)
-   - **Quality**: Still selects high-quality representatives based on length
-     and quality
```
Comment (evanfields) [inline README.md]: Why don't we want to describe the choice of Python algorithm options anymore?

File: README.md [hunk 9]
```diff
@@ -87,102 +89,233 @@ dedup_params = DedupParams(
...
-# Configure minimizer parameters
-# Note: For large datasets with many reads, use a larger kmer_len to avoid
-# bucket explosions. With default kmer_len=7, there are only 4^7 (~16k)
-# possible sequences.
+# Configure minimizer parameters (rarely needs changing)
 minimizer_params = MinimizerParams(
     num_windows=3,    # Number of windows per read
     window_len=25,    # Base pairs per window
-    kmer_len=7        # K-mer size (use 15 for large datasets)
+    kmer_len=7        # K-mer size for minimizers
```
Comment (evanfields) [inline README.md]: Since the graph and streaming algorithms bucket differently, the ideal kmer size (possibly other minimizer parameters) seems different. Should we have defaults by algorithm? That could also allow the Python default and Rust default to more neatly match.

(Feel free to punt to an issue, even if you agree.)

File: README.md [hunk 10]
```diff
@@ -87,102 +89,233 @@ dedup_params = DedupParams(
...
+- For each cluster, an exemplar is selected based on graph centrality, quality
+  score, read length, and read ID
 
-During Pass 1, when a better representative is found for a cluster, earlier reads
-in that cluster still point to the old representative. Pass 2 resolves this by
-looking up the final best representative for each cluster using a simple dictionary
-lookup (no re-bucketing or re-matching needed).
+**Rust and Python streaming**:
+- No graph construction - purely streaming approach
+- First matching read in a bucket becomes the cluster exemplar
```
Comment (evanfields) [inline README.md]: optional nit: I suggest rephrasing here; the current phrasing makes it seem like the first read in a cluster is always the cluster exemplar, which isn't true. Maybe something about how the first read "identifies" a cluster and the exemplar is the best read (based on quality and length, though see a relevant comment in this review) of all those that match the initial cluster-identifying read?

Comment (evanfields) [reply README.md]: Also applies a few lines down (264 in this version)

File: tests/test_dedup.py [hunk 11]
```diff
@@ -38,44 +39,32 @@ def _random_seq(length: int, rng: random.Random) -> str:
 class TestHelperFunctions:
     """Test sequence manipulation helper functions."""
 
-    @pytest.mark.fast
-    @pytest.mark.unit
```
Comment (evanfields) [inline tests/test_dedup.py]: Why don't we want to annotate tests anymore?

File: tests/test_dedup.py [hunk 12]
```diff
@@ -620,3 +540,289 @@ def test_realistic_reads_with_errors(self, dedup_func):
...
+    subprocess.run(
+        ["maturin", "develop", "--quiet"],
+        cwd=str(rust_bindings_dir),
+        check=True
+    )
+    from nao_dedup_rust import deduplicate_read_pairs_rust
+except (subprocess.CalledProcessError, FileNotFoundError) as e:
+    raise RuntimeError(
+        "Failed to build Rust library. "
+        "Install Rust first (brew install rust), then:\n"
```
Comment (evanfields) [inline tests/test_dedup.py]: optional nit: `brew install rust` isn't platform agnostic, maybe just refer to the readme?

File: tests/test_dedup.py [hunk 13]
```diff
@@ -620,3 +540,289 @@ def test_realistic_reads_with_errors(self, dedup_func):
...
+        mapping = _get_exemplar_mapping(result)
+        assert mapping == {}
+
+    def test_single_read(self, dedup_func):
+        rp = ReadPair("read1", "AAAA", "TTTT", "IIII", "IIII")
+        result = dedup_func([rp], verbose=False)
+        mapping = _get_exemplar_mapping(result)
+        assert mapping == {"read1": "read1"}
+
+    def test_identical_reads(self, dedup_func):
```
Comment (evanfields) [inline tests/test_dedup.py]: Some of these tests exactly duplicate (except for their parameterization) the prior Python-only end-to-end dedup tests. E.g. see `test_all_identical_sequences_single_cluster` in this file.

Just confirming you prefer to keep them separate so the tests are organized as two sets of pairwise comparisons:
* Python graph vs Python streaming
* Python streaming vs Rust streaming

Rather than just having one set of end-to-end tests that we parameterize to hit all 3 implementations?

File: src/lib.rs [hunk 14]
```diff
@@ -0,0 +1,420 @@
...
+
+fn extract_minimizers(seq: &str, params: &MinimizerParams) -> Vec<u64> {
+    let seq_bytes = seq.as_bytes();
+    let seq_len = seq_bytes.len();
+
+    if seq_len < params.kmer_len {
+        return vec![];
+    }
+
+    assert!(params.kmer_len <= 32, "k-mer length must be <= 32 for 2-bit encoding");
```
Comment (evanfields) [inline src/lib.rs]: We want an assert not a hard error here?

File: src/lib.rs [hunk 15]
```diff
@@ -0,0 +1,420 @@
...
+        all_mins.extend(rev_mins);
+
+        // Find matching exemplar by checking buckets (use FIRST match found)
+        let mut matching_exemplar: Option<String> = None;
+
+        'outer: for &min_hash in &all_mins {
+            if let Some(bucket_reads) = self.buckets.get(&min_hash) {
+                for candidate_id in bucket_reads {
+                    // Look up candidate in exemplar_store (not all reads)
+                    if let Some(candidate) = self.exemplar_store.get(candidate_id) {
```
Comment (evanfields) [inline src/lib.rs]: The Python implementation accumulates a set of checked IDs to avoid double-checking some exemplars (that share multiple keys with our read pair). We don't want that in this implementation?

File: src/lib.rs [hunk 16]
```diff
@@ -0,0 +1,420 @@
...
+                }
+            }
+        }
+
+        let exemplar_id = if let Some(exemplar) = matching_exemplar {
+            // Found a match - add to existing cluster
+            if let Some(cluster) = self.clusters.get_mut(&exemplar) {
+                cluster.count += 1;
+                // Update best read if this one is better
+                if mean_q > cluster.best_score {
```
Comment (evanfields) [inline src/lib.rs]: Semantic difference from Python implementation: this is choosing a cluster best-so-far based on just mean quality, whereas the Python side uses a quality-length weighted combination.

File: src/lib.rs [hunk 17]
```diff
@@ -0,0 +1,420 @@
...
+    } else {
+        (1u64 << (2 * params.kmer_len)) - 1
+    };
+
+    let mut minimizers = Vec::with_capacity(params.num_windows);
+    let step = if params.num_windows > 1 {
+        (seq_len.saturating_sub(params.window_len)) / (params.num_windows - 1)
+    } else {
+        0
+    };
```
Comment (evanfields) [inline src/lib.rs]: Meaningful difference from Python implementation: Python takes adjacent windows at the start of the read; this is spacing the windows across the read.

Is this difference intentional?

**Review by evanfields — COMMENTED** [review]
Thanks for the Rust implementation! I learned a lot while reading this PR.

Apologies for so many comments. Many are very minor, but I would especially like to hear about the semantic differences between the Rust and Python implementations.

File: README.md [hunk 18]
```diff
@@ -87,102 +89,233 @@ dedup_params = DedupParams(
...
-2. **Bucketing**: Read pairs with matching minimizers are assigned to the same
-   buckets, reducing the number of pairwise comparisons needed.
+#### K-mer Hashing
 
-3. **Pairwise Comparison**: Within each bucket, read pairs are compared to
-   determine if they're duplicates, allowing for:
-   - Small alignment offsets (configurable via `max_offset`)
-   - Sequencing errors (configurable via `max_error_frac`)
-   - Optional mate-pair orientation swaps (configurable via `orientation`)
+K-mers are hashed using a 2-bit DNA encoding (A=0, C=1, G=2, T=3) that encodes
```
[the remark this answers is not in the document]
Comment (jeffkaufman) [reply README.md]: This is a difference between programmer-speak and CS-speak.  To a programmer, a hash is converting something into a good key for a hash table, minimizer, or other algorithm that needs well-distributed numbers.  To a CS person, though, a hash fundamentally requires losing information which this doesn't.

Added comment in c7c466a and 7cb1f4a

File: src/lib.rs (hunk 14 above)
Comment (jeffkaufman) [reply src/lib.rs]: Good point: since this is not a programmer error it shouldn't be an assert.  Fixed in b6df39a

File: src/lib.rs [hunk 19]
```diff
@@ -0,0 +1,470 @@
...
+
+            let final_cluster_id = if let Some(cluster) = self.clusters.get(&cluster_key) {
+                cluster.best_read_id.clone()
+            } else {
+                cluster_key
+            };
+
+            final_results.insert(read_id.clone(), final_cluster_id);
+        }
+
```
[the remark this answers is not in the document]
Comment (jeffkaufman) [reply src/lib.rs]: Fixed in 1cf4415

File: tests/test_dedup.py (hunk 11 above)
Comment (jeffkaufman) [reply tests/test_dedup.py]: Yes, let's add this later if we need it

File: Cargo.toml [hunk 20]
```diff
@@ -0,0 +1,13 @@
+[package]
+name = "nao_dedup"
+version = "0.1.0"
+edition = "2024"
```
[the remark this answers is not in the document]
Comment (jeffkaufman) [reply Cargo.toml]: Decided not to do this

File: tests/test_dedup.py [hunk 21]
```diff
@@ -826,3 +752,213 @@ def test_cluster_lookup_multiple_updates(self, dedup_func):
...
+            Reverse read: beginning of top strand (reported as-is by sequencer)
+
+        This means: (F_alpha, R_alpha) should match (R_beta, F_beta)
+        with NO reverse complement needed.
+
+        Example with GATTACA insert (using longer sequences for realistic minimizers):
+        """
+        # Create a 150bp "insert" - this represents the actual DNA fragment
+        # Use a pattern that's clearly directional (not palindromic)
+        insert_top = "GATTACA" * 20 + "ACATAG" * 5  # 150bp total
```
Comment (evanfields) [inline tests/test_dedup.py]: nit: GATTACA is 7, ACATAG is 6, `7 * 20 + 6 * 5 = 170 != 150`

File: src/lib.rs [hunk 22]
```diff
@@ -73,75 +97,135 @@ impl ReadPair {
...
+const ENCODE_LOOKUP: [u64; 256] = {
+    let mut table = [u64::MAX; 256];
+    table[b'A' as usize] = 0;
+    table[b'a' as usize] = 0;
+    table[b'C' as usize] = 1;
+    table[b'c' as usize] = 1;
+    table[b'G' as usize] = 2;
+    table[b'g' as usize] = 2;
+    table[b'T' as usize] = 3;
+    table[b't' as usize] = 3;
```
Comment (evanfields) [inline src/lib.rs]: `u/U` as well?

**Review by evanfields — APPROVED** [review]
Tiny comments, no need to re-review, exciting stuff. Thanks for all your work here!

File: README.md [hunk 23]
```diff
@@ -223,11 +249,39 @@ each base into exactly 2 bits. This provides several advantages:
...
+Windows are placed adjacently starting from the beginning of each read (positions 0, window_len, 2×window_len, etc.). This strategy:
+- Aligns with standard NGS duplicate detection based on 5' mapping coordinates
+- Focuses on the most stable region of reads (the beginning)
+- Avoids the tail region, which is most likely to be trimmed or contain sequencing errors
+
+For example, with 3 windows of 25bp on a 150bp read:
+- Window 0: positions [0-25]
+- Window 1: positions [25-50]
+- Window 2: positions [50-75]
+- Positions 75-150 are not examined for minimizers
```
[the remark this answers is not in the document]
Comment (jeffkaufman) [reply README.md]: done in 4d3c9f0

File: tests/test_dedup.py [hunk 24]
```diff
@@ -826,3 +752,213 @@ def test_cluster_lookup_multiple_updates(self, dedup_func):
...
+        assert mapping["readB"] == "readB"
+
+    def test_windows_with_all_ns(self, dedup_func):
+        """
+        Test that sequences with windows containing all N's are handled correctly.
+
+        Edge case: When a window has all N's (or enough N's that no valid k-mer
+        can be formed), Python's _extract_minimizer returns EMPTY_KMER_SENTINEL_HASH
+        while Rust's extract_minimizers skips the window (doesn't add to vector).
+        Both should produce the same clustering behavior.
```
[the remark this answers is not in the document]
Comment (jeffkaufman) [reply tests/test_dedup.py]: Commented in 7fa6e96

File: tests/test_dedup.py (hunk 21 above)
Comment (jeffkaufman) [reply tests/test_dedup.py]: Fixed in ba25f49
